src/states/elder/mod.rs

```diff
@@ -665,178 +656,151 @@ impl Elder {
             );
             return Err(RoutingError::InvalidSource);
         }
 
-        if let Some(signed_msg) = self.sig_accumulator.add_proof(msg) {
-            let signed_msg = HopMessageWithBytes::new(signed_msg, &self.log_ident())?;
-            self.handle_signed_message(signed_msg)?;
+        if let Some(msg) = self.sig_accumulator.add_proof(msg) {
+            self.handle_accumulated_message(msg)?
         }
-        Ok(())
+
+        Ok(Transition::Stay)
     }
 
-    // If the message is for us, verify it then, handle the enclosed routing message and swarm it
-    // to the rest of our section when destination is targeting multiple; if not, forward it.
-    fn handle_signed_message(&mut self, msg: HopMessageWithBytes) -> Result<(), RoutingError> {
-        if !self.routing_msg_filter.filter_incoming(&msg).is_new() {
+    fn handle_accumulated_message(&mut self, msg: Message) -> Result<()> {
+        // FIXME: this is almost the same as `Base::handle_new_deserialized_message` - find a way
+        // to avoid the duplication.
+
+        let mut msg = MessageWithBytes::new(msg, &self.log_ident())?;
+
+        if !self.filter_incoming_message(&msg) {
             trace!(
                 "{} Known message: {:?} - not handling further",
                 self,
-                msg.full_message_crypto_hash()
+                msg.full_crypto_hash()
             );
+
             return Ok(());
         }
 
-        self.handle_filtered_signed_message(msg)
-    }
-
-    fn handle_filtered_signed_message(
-        &mut self,
-        mut msg: HopMessageWithBytes,
-    ) -> Result<(), RoutingError> {
-        trace!(
-            "{} - Handle signed message: {:?}",
-            self,
-            msg.full_message_crypto_hash()
-        );
-
-        let in_location = self.in_dst_location(msg.message_dst());
-        if !in_location || msg.message_dst().is_multiple() {
-            // Relay closer to the destination or broadcast to the rest of our section.
-            if let Err(error) = self.send_signed_message(&msg) {
-                debug!(
-                    "{} Failed to send {:?}: {:?}",
-                    self,
-                    msg.full_message_crypto_hash(),
-                    error
-                );
-            }
+        if self.should_relay_message(msg.message_dst()) {
+            self.relay_message(None, msg.clone())?;
         }
 
-        if in_location {
-            let signed_msg = msg.take_or_deserialize_signed_routing_message()?;
-            let handle = self.verify_signed_message(&signed_msg).map_err(|error| {
-                self.log_verify_failure(&signed_msg, &error, self.chain.get_their_key_infos());
-                error
-            })?;
-
-            if handle {
-                self.update_our_knowledge(&signed_msg);
-                self.routing_msg_queue.push_back(signed_msg);
-            }
+        let msg = msg.take_or_deserialize_message()?;
+        if self.should_handle_message(&msg)? {
```

Comment:
Same `take_or_deserialize_message` should only happen after or within should_handle_message whne we know we need it.